Initial Commit of VirtFS-XFS

jerryk55 opened the pull request:
Initial integration of XFS with VirtFS.
Module and class name modifications.
Addition of file_class and dir_class.

@roliveri @hsong-rh @movitto Can be looked at but not ready for full review yet.

jerryk55 committed:
Initial Commit of VirtFS-XFS

Initial integration of XFS with VirtFS.
Module and class name modifications.
Addition of file_class and dir_class.

jerryk55 commented:
@miq-bot add_label wip

Fryguy commented on lib/virtfs/xfs.rb:
```diff
@@ -1 +1,20 @@
+
+require_relative "xfs/base"
+require_relative "xfs/file_class"
+require_relative "xfs/dir_class"
```
Minor, but it's common in gems to just refer to file assuming lib is on the path.  That is, instead of `require_relative 'xfs/whatever'`, prefer `require 'virtfs/xfs/whatever'`

jerryk55 replied:
@Fryguy I have no problem with your suggestion other than the fact that the VirtFS gem written by @roliveri and three PRs for gems VirtFS-NTFS, VirtFS-Ext3, and VirtFS-Ext4 by @movitto do it this way - I just used the same pattern.  We should either change them all or keep them all this way.

roliveri replied:
@Fryguy Is there any benefit in using `require` instead of `require_relative`? Shouldn't `require_relative`  be more efficient, since it doesn't have to go through the array of load paths? It seems to me, GEM-internal requires like these are ideal candidates for the use of `require_relative`.

Fryguy replied:
Neither are more or less efficient.  This is a standard pattern in every gem.

roliveri replied:
@Fryguy I don't see how that can be, since one depends on the size of load path array and the other does not. I can see that the difference may  be negligible, and thus, not a factor.

From what I've seen, it's not the standard pattern for every gem. I've seen a number of gems that use require_relative, mostly newer ones. True, most gems don't use it, but that's probably because require_relative hasn't been around as long. Unless there's some downside in using require_relative, I don't see any reason not to use it here.

Fryguy replied:
>  I don't see how that can be, since one depends on the size of load path array and the other does not. I can see that the difference may be negligible, and thus, not a factor.

That's fair, but yeah, it's negligible is more what I meant.  Once the file is detected, they both get path expanded and put into $LOADED_FEATURES (sort of...they are actually put into an internal Hash for efficient lookup).

Fryguy replied:
http://guides.rubygems.org/patterns/#loading-code

rubygems.org pattern describes using both, so I am fine with it.  From my experience most gems load relative to the $LOAD_PATH (which lib is put on the $LOAD_PATH).

jerryk55 replied:
Ok - so `require_relative` is ok.

Fryguy commented on lib/virtfs/xfs/directory3_data_header.rb:
```diff
@@ -1,7 +1,7 @@
-require 'fs/xfs/directory'
-require 'fs/xfs/superblock'
+require_relative 'directory'
+require_relative 'superblock'
```
These lines should not be needed at all if the gem is loaded properly (i.e. they are loaded in virtfs/xfs.rb), so they can be remove entirely

Fryguy commented on lib/virtfs/xfs/file_class.rb:
```diff
@@ -0,0 +1,224 @@
+module VirtFS::XFS
+  class FS
```
File name is `file_class.rb` but the class is `FS`.  They should match.

jerryk55 replied:
Agreed.  Also based on previous PRs for the other retypes.  @movitto I will make this comment on your PRs as well.

roliveri replied:
@jerryk55 As per my email. The FS class is composed of a number of files, containing the implementations of the class and instance methods of the File and Dir classes - all of which are instance methods of the FS. Instead of lumping them in one big file, I placed them in files based on those categories. To be consistent with my original prototype however, this file should probably be named `fs_file_class.rb`.

jerryk55 replied:
@roliveri I take it you meant `xfs_file_class.rb`

roliveri replied:
@jerryk55 Yes, but I guess `fs` would work as well, since the class name is `FS`. Actually, the camcorderfs uses `fs` in the file names instead of `camcorderfs` or something like that. However, that was done to more easily distinguish the top and bottom layers of the recording boundary.

jerryk55 replied:
@roliveri but your protofs and nativefs code under the VirtFS use the full pathname.  So we need to pick one and be consistent.

roliveri replied:
@jerryk55 when you say "pathname" I assume you meant "filesystem"?

Consistency is nice, but sometimes it's not possible given the underlying needs of some implementations. We can't dictate the class structure of the FS modules, because sometimes that structure won't apply to the implementation.

For example: NativeFS implements both thick and thin interfaces. Because of that, there is no "FS" class or NativeFS class, there's a "Thin" class and "Thick" class.

Another example is CamcorderFS. Here we have an "FS" class because the complexity of the implementation requires code under the CamcorderFS namespace that's not part of the exported filesystem.

All we can really dictate is that the module implementation needs to implement a "filesystem" class and what that class needs to provide.

As far as filenames go, maybe they should be consistent with the implementation, not between modules. The alternative could be that we just say that all files that define parts of the exported filesystem class start with `fs_`. However, for modules that only implement the filesystem class, that could seem redundant - in that all the files would start with `fs_`.

Fryguy replied:
I don't really understand what you mean.  Whether the FS class exists or not doesn't pertain to the expectation that class and module names to match their respective file path.  This is standard in every gem.

So, I see in this PR there are multiple files that implement parts of FS.  This is confusing for people used to the gem naming conventions, and it will take them a while to understand where methods are defined when mutiple files are reopening the same class.  This can easily be remedied using normal Ruby mechanisms.  

For example, right now lib/virtfs/xfs/base.rb, lib/virtfs/xfs/file_class.rb, and lib/virtfs/xfs/dir_class.rb all implement parts of the FS class.  As an external contributor I would expect it to all be in lib/virtfs/xfs/fs.rb.  Understanding that you need to split it up for clarity, I would expect the following (roughly...if you need to mixin the class methods, you can do so in a couple ways, which are pretty straightforward):

```ruby
# Tree
# lib
# \-virtfs
#   \-xfs
#     \-fs
#     | |-base.rb
#     | |-dir_methods.rb
#     | \-file_methods.rb
#     \-fs.rb

# lib/virtfs/xfs/fs.rb:
module VirtFS::XFS
  class FS
    include FS::Base
    include FS::FileMethods
    include FS::DirMethods
  end
end

# lib/virtfs/xfs/fs/base.rb:
module VirtFS::XFS
  class FS
    module Base
      # The current base.rb file
    end
  end
end

# lib/virtfs/xfs/fs/file_methods.rb:
module VirtFS::XFS
  class FS
    module FileMethods
      # The current file_class.rb file
    end
  end
end

# lib/virtfs/xfs/fs/dir_methods.rb:
module VirtFS::XFS
  class FS
    module DirMethods
      # The current dir_class.rb file
    end
  end
end
```

jerryk55 replied:
@roliveri the suggestion above by @Fryguy above makes perfect sense.  Is this an ok way to go from your perspective?

roliveri replied:
@Fryguy Actually, I tried the mixin route initially and I found it to be more cumbersome overall.
Also, I recall I may have encountered some technical snags as well. But that may have been in the core VirtFS code - which has the same issue.

I'm not opposed to the change. FS modules are far simpler than the core VirtFS, so it may not be an issue. There are three areas that might be problematic: the core VirtFS, the NativeFS (thick vs thin) and CamcorderFS (which is the most complex FS module).

So, I say we go with the mixin's for the simple FS modules, and then revisit the more complex cases to see if we can address those as well.

Fryguy commented:
There is a lot of mixture of cleanup, new files, refactoring all in one commit, which makes review really hard...curious if you can avoid that with multiple commits.

jerryk55 committed:
Review Comments for Initial Commit

The initial review of this PR concentrated on two areas that have been updated:
1) Whether to use require_relative or require, as well as not needing to use
   unnecessary requires.
2) File naming conventions for classes spanning multiple source files, now
   using a mixin file to merge the FS class from multiple files with different
   modules.

jerryk55 commented:
Ok - as per lengthy review discussion above, unnecessary requires / require_relatives removed, and the FS class files have been renamed and a mixing used to merge the multiple modules into the class.  Still WIP at this point.

jerryk55 committed:
Remove illformed requirement from Gemspec

An attempt to add a github requirement to the master branch of the VirtFS gem was
illformed and is removed here.

roliveri commented on lib/virtfs/xfs/allocation_group.rb:
```diff
@@ -1,12 +1,11 @@
+module VirtFS::XFS
   # ////////////////////////////////////////////////////////////////////////////
 
   AG_FREESPACE = BinaryStruct.new([
```
I assume all of these constants are specific to `AllocationGroup`, given they're defined in this file. Yet the constants are defined under the `VirtFS::XFS` namespace. It would probably be better to define them under `VirtFS::XFS::AllocationGroup`. Then, there would also be no need for the `AG_` prefix.

This is a general comment. I've noticed that in many class-specific files, you define constants directly under the `VirtFS::XFS` namespace, instead of under the class in question.

jerryk55 replied:
Agreed.  Legacy from the original implementation of the filesystem, but regardless this makes more sense with the modifications.  Across all classes in the filesystem where appropriate.

roliveri commented on lib/virtfs/xfs/base.rb:
```diff
@@ -0,0 +1,51 @@
+module VirtFS::XFS
+  class FS
+    module Base
```
Given the naming convention, the `Base` module should be in the file: `lib/virtfs/xfs/fs/base.rb`.

jerryk55 replied:
Done.

roliveri commented on lib/virtfs/xfs/bmap_btree_block.rb:
```diff
@@ -1,10 +1,10 @@
 
-module XFS
+module VirtFS::XFS
   # ////////////////////////////////////////////////////////////////////////////
```
Same comment as above about defining class-specific constants directly under `VirtFS::XFS`.

jerryk55 replied:
Done.

roliveri commented on lib/virtfs/xfs/bmap_btree_record.rb:
```diff
@@ -1,4 +1,4 @@
-module XFS
+module VirtFS::XFS
   BMAP_BTREE_REC = BinaryStruct.new([
```
Same comment about constants under `VirtFS::XFS`.

roliveri commented on lib/virtfs/xfs/bmap_btree_root_node.rb:
```diff
@@ -1,6 +1,4 @@
-require 'fs/xfs/inode'
-
-module XFS
+module VirtFS::XFS
```
Same comment about constants under `VirtFS::XFS`.

roliveri commented on lib/virtfs/xfs/dir_methods.rb:
```diff
@@ -0,0 +1,79 @@
+module VirtFS::XFS
+  class FS
+    module DirMethods
```
This module should be in the file: `lib/virtfs/xfs/fs/dir_methods.

Also, I've been naming this module `DirClassMethods`, given it doesn't define instance methods. I also include a comment explaining that Dir class methods are implemented as instance methods of the filesystem.

jerryk55 replied:
Yep.

roliveri commented on lib/virtfs/xfs/directory.rb:
```diff
@@ -1,11 +1,4 @@
-require 'fs/xfs/short_form_directory_entry'
-
-module XFS
+module VirtFS::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/directory2_data_header.rb:
```diff
@@ -1,7 +1,4 @@
-require 'fs/xfs/superblock'
-
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/directory3_data_header.rb:
```diff
@@ -1,7 +1,4 @@
-require 'fs/xfs/superblock'
-
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/directory_block_tail.rb:
```diff
@@ -1,4 +1,4 @@
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/directory_data_header.rb:
```diff
@@ -1,9 +1,4 @@
-require 'fs/xfs/superblock'
-
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/directory_entry.rb:
```diff
@@ -1,6 +1,6 @@
 require 'binary_struct'
 
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/file_methods.rb:
```diff
@@ -0,0 +1,226 @@
+module VirtFS::XFS
+  class FS
+    module FileMethods
```
This class should be in file: `lib/virtfs/xfs/fs/file_methods.rb`.

Also, as I mentioned for `dir_methods`, I've been naming this `FileClassMethods`.

roliveri commented on lib/virtfs/xfs/inode.rb:
```diff
@@ -1,12 +1,8 @@
-require 'fs/xfs/bmap_btree_root_node'
 
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/short_form_directory_entry.rb:
```diff
@@ -1,7 +1,6 @@
-require 'fs/xfs/inode'
 
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/short_form_header.rb:
```diff
@@ -1,4 +1,4 @@
-module XFS
+module Virtfs::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/superblock.rb:
```diff
@@ -1,16 +1,13 @@
 require 'rufus/lru'
 
-module XFS
+module VirtFS::XFS
```
Same comment about constants under VirtFS::XFS.

roliveri commented on lib/virtfs/xfs/allocation_group.rb:
```diff
@@ -1,12 +1,11 @@
 require 'binary_struct'
-require 'util/miq-uuid'
+require 'virt_disk/disk_uuid'
```
Where is this file?

roliveri commented on lib/virtfs/xfs/file_methods.rb:
```diff
@@ -0,0 +1,226 @@
+      def file_lstat(p)
+        de = get_file(p)
+        raise Errno::ENOENT, "No such file or directory #{p}" if de.nil?
+        VirtFS::Stat.new(VirtFS::XFS::File.new(de, superblock).to_h)
```
Where else is this `File` class used? If this is the only place, I think having a `File` class used only for this purpose is confusing. Maybe you should just replace it with a method that returns the hash.

roliveri commented on lib/virtfs/xfs/file.rb:
```diff
@@ -0,0 +1,14 @@
+module VirtFS::XFS
+  class File
```
See my comment in `file_lstat`, about this `File` class.

jerryk55 replied:
Just a note - should I assume you've made this same comment, perhaps offline, to @movitto regarding https://github.com/ManageIQ/virtfs-ext3/pull/1 ? This code was modified from that PR.

roliveri commented on lib/virtfs/xfs/file_methods.rb:
```diff
@@ -0,0 +1,226 @@
+      def file_new(p, parsed_args, _open_path, _cwd)
+        de = get_file(p)
+        raise Errno::ENOENT, "No such file or directory #{p}" if de.nil?
+        de
```
The `file_new` method needs to return an object that implements the minimum methods required for the File instance thin interface. Does this directory entry object do that? It doesn't seem so, given the `file_size` method calls `superblock.inode(de.inode).length` instead of just calling `de.size`.

The object returned by this method is wrapped by the `ThinFileDelegator` of `VirtFS`. In the `ThinFileDelegator` object, it is accessed as `@fs_io_obj` and needs to implement all the methods the `ThinFileDelegator` expects of it.

jerryk55 replied:
@roliveri Should we assume that you've commented on @movitto's https://github.com/ManageIQ/virtfs-ext3/pull/1 PR since this implementation was glommed from there?

jerryk55 committed:
Move constants from VirtFS::XFS namespace to VirtFS::XFS::*

As per review comments, based on legacy XFS code that was
migrated into this repo, some constants previously lived
under VirtFS::XFS.  These have been moved into the appropriate
classes in each file below as necessary.

jerryk55 committed:
Rename and Move FS Class Files Per Review Comments

As per review comments, move Base, DirMethods, and FileMethods class
source files to the "fs" subdirectory.  Rename DirMethods and
FileMethods to DirClassMethods and FileClassMethods.

jerryk55 committed:
Add virtfs gem dependency to Gemfile

jerryk55 committed:
Use UUIDTools where necessary

Prior version of the XFS code used MIQUUID which is no longer relevant in this gem.
The previous commit of these files attempted to use a non-existent virt_disk uuid
class.  Now changed to use UUIDTools::UUID.

jerryk55 commented:
Just a note - all of @roliveri comments addressed except the File class and the file_new method.  I wanted to push out this set of commits and will address the remainder tomorrow.

jerryk55 committed:
Fix Capitalization of VirtFS Class

jerryk55 committed:
Fix Requires for VirtFS::XFS::FS Class Files

jerryk55 committed:
Remove VirtFS::XFS::File Class

The File class named above is really not required.
The only relevant method has been moved into the FileClassMethods class
and will return a hash for the directory entry.

jerryk55 committed:
Old Review Comments

Added code implementing changes based on old review from the end of 2016.
Major changes Included are those relating to file operations.
Minor changes around spacing and constant naming.

jerryk55 committed:
Add binary_struct gem dependency

We need the binary_struct gem dependency for our spec tests.

jerryk55 committed:
Gem changes for initial spec

Add other gems for initial spec test.

jerryk55 commented:
@roliveri @hsong-rh I finally went back and addressed the *year-old*! review comments on this initial commit of the virtfs-xfs repo, and updated the gemspec to allow the initial dummy spec test to pass.  We can merge this so it is out in the real world and if we want to look at adding more detailed tests to all the virtfs repos as followup PRs we can do that.
